official/coin_build/man/man3/SoFont.3
.TH "SoFont" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoFont \- The \fBSoFont\fP class is an appearance node for setting fonts\&.
.PP
Successive text rendering nodes (like \fBSoText2\fP, \fBSoText3\fP, \fBSoAsciiText\fP, etc) will use the font specified from an \fBSoFont\fP node when visualizing text\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/nodes/SoFont\&.h>\fP
.PP
Inherits \fBSoNode\fP\&.
.PP
Inherited by \fBSoFontStyle\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "\fBSoFont\fP (void)"
.br
.ti -1c
.RI "virtual void \fBdoAction\fP (\fBSoAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBgetBoundingBox\fP (\fBSoGetBoundingBoxAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBGLRender\fP (\fBSoGLRenderAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBcallback\fP (\fBSoCallbackAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBpick\fP (\fBSoPickAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBgetPrimitiveCount\fP (\fBSoGetPrimitiveCountAction\fP *action)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBSoSFName\fP \fBname\fP"
.br
.ti -1c
.RI "\fBSoSFFloat\fP \fBsize\fP"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "virtual const \fBSoFieldData\fP * \fBgetFieldData\fP (void) const"
.br
.ti -1c
.RI "virtual \fB~SoFont\fP ()"
.br
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static const \fBSoFieldData\fP ** \fBgetFieldDataPtr\fP (void)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
The \fBSoFont\fP class is an appearance node for setting fonts\&.
.PP
Successive text rendering nodes (like \fBSoText2\fP, \fBSoText3\fP, \fBSoAsciiText\fP, etc) will use the font specified from an \fBSoFont\fP node when visualizing text\&. 

The font name is recognized with the form 'family:style'\&. The range of supported font families depends on which fonts are installed on the system\&. A typical font family might be 'Arial' or 'Times New
Roman'\&.
.PP
Style is either 'Bold', 'Italic' or 'Bold Italic'\&.
.PP
Spaces before or after the ':' will be ignored\&.
.PP
If a good match for the chosen font cannot be found, the default font will be loaded instead\&. The default 2D font is a builtin 8x12 points font\&. The 3D font is a serif font ala Times New Roman\&. It is not possible to apply styles to any of the builtin fonts\&. It is not possible to specify a size for the default 2D font\&.
.PP
One can explicitly select the default font by setting 'defaultFont' as the font name, or by just leaving the \fBSoFont::name\fP field alone, as it has that value by default\&. In that regard, note that \fBSoFont\fP does not inherit / accumulate any parts from earlier SoFont-nodes, even when one or both of its fields are left blank in e\&.g\&. an iv-file, like this:
.PP
.PP
.nf
#Inventor V2.1 ascii

Font { name "Helvetica"  size 50 }
Text2 { string "hepp" }

Translation { translation 5 80 0 }

BaseColor { rgb 1 1 0 }

Font { size 30 }
Text2 { string "svupp" }
.fi
.PP
.PP
(The second string in the above example will be rendered with the default built-in font, not in Helvetica\&.)
.PP
The default fonts are always accessible by Coin as they are embedded into the run-time library\&. If one needs to guarantee that the text will have the same appearance under all circumstances, the default font will be a safe choice\&. Another solution might be to use the \fIFreeType\fP font engine and explicitly name a font-file\&. This is described in more detail below\&.
.PP
Here is a simple example on how to print a string using a bold and italic Arial font:
.PP
.PP
.nf
#Inventor V2.1 ascii

Separator {
   Font {
     name "Arial:Bold Italic"
     size 14
   }
   Text2 {
     string ["This is a", "Coin font test"]
     justification CENTER
   }    
}
.fi
.PP
.PP
Coin has support for two different font APIs\&. On non-Windows platforms, the \fIFreeType\fP library together with the optional \fIFontconfig\fP library is used\&.
.PP
If the \fIFontconfig\fP library is installed the font file on the sytem for a given font name will be located through it\&. For more information on \fIFontconfig\fP see http://freedesktop.org/software/fontconfig \&. Additionally \fIFontconfig\fP allows to match specific fonts through its own pattern matching format (see http://pdx.freedesktop.org/~fontconfig/fontconfig-user.html)\&. Please note that the point size value in the textual representation of the Fontconfig pattern is currently overriden by the size field of the \fBSoFont\fP node and if no size field is specified the default value of the size field is in effect\&. In case you intend to use your application on systems where the \fIFontconfig\fP library is expected to be not installed you should not make use of \fIFontconfig's\fP font pattern syntax\&. \fIFontconfig\fP usage can be prevented by setting the 'COIN_FORCE_FONTCONFIG_OFF' environment variable to 1\&.
.PP
Here is an example on how to print a string using the \fIFontconfig\fP pattern matching syntax using a bold & italic font where Times New Roman is the preferred font family\&.
.PP
.PP
.nf
#Inventor V2.1 ascii

Separator {
   Font {
     name "Times New Roman,Arial,Helvetica:italic:bold"
     size 24
   }
   Text2 {
     string ["This is a", "Coin font test"]
     justification CENTER
   }    
}
.fi
.PP
.PP
On Windows the Win32 GDI library is used\&. \fIFreeType\fP and \fIFontconfig\fP are dynamically loaded on demand by Coin if font support are requested by a node\&. When fontsupport is loaded on Windows, FreeType will have precedence over Win32 if located\&. This can be prevented by setting the 'COIN_FORCE_FREETYPE_OFF' environment variable to 1\&. When using FreeType, you need FreeType version 2\&.1 or later\&. On Mac OS X, version 2\&.1\&.7 or later is required\&.
.PP
If Coin cannot load the \fIFreeType\fP library, and is not running on Microsoft Windows, only the default fonts will be accessible\&.
.PP
It is possible to specify the TrueType font file directly if \fIFreeType\fP is used as the font engine\&. This is done by including the '\&.ttf' in the filename, i\&.e\&. 'Comic_Sans_MS\&.ttf'\&. Coin will then search the local path for the running program and then the path specified by the 'COIN_FONT_PATH' environment variable\&. If the program is using \fIFreeType\fP on a Windows platform, the '$WINDIR/Fonts' directory will also be searched\&.
.PP
It is not possible to directly specify a TrueType font file if Windows is handling the fonts\&. This is due to the way Windows is accessing the fonts through the system registry\&. All fonts must therefore be properly installed and given a system name\&. Open the 'Control Panel' and double click on the 'Fonts' icon for an overview of installed fonts and their names\&.
.PP
Beware that some non-English versions of Windows are using different name for the styles (i\&.e\&. 'Italique' instead of 'Italic')\&. These names are supported in Coin, but it is recommended for portability purposes to only use the English terms\&. Please note that there is still a possibility that there are no fonts installed using the terms 'Bold' or 'Italic' on the Windows platform\&. To guarantee that a font is accessible you must either use the \fIFreeType\fP library and include a TrueType font in your distribution, or you must avoid using styles and stick to the standard Windows fonts\&.
.PP
If the 'COIN_DEBUG_FONTSUPPORT' environment variable is set to 1, an extensive amount of information about loading, initializing and using fonts will be output\&. Issues like missing fonts and other related problems will then be reported, so we advice you to first try to use that debugging option when something does not work quite as expected\&.
.PP
\fBFILE FORMAT/DEFAULTS:\fP 
.PP
.nf
Font {
    name "defaultFont"
    size 10
}

.fi
.PP
.PP
\fBSee also:\fP
.RS 4
\fBSoFontStyle\fP, \fBSoText2\fP, \fBSoText3\fP, \fBSoAsciiText\fP 
.RE
.PP

.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoFont::SoFont (void)"
Constructor\&. 
.SS "SoFont::~SoFont ()\fC [protected]\fP, \fC [virtual]\fP"
Destructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "\fBSoType\fP SoFont::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. Usage example:
.PP
.PP
.nf
void foo(SoNode * node)
{
  if (node->getTypeId() == SoFile::getClassTypeId()) {
    SoFile * filenode = (SoFile *)node;  // safe downward cast, knows the type
  }
}
.fi
.PP
.PP
For application programmers wanting to extend the library with new nodes, engines, nodekits, draggers or others: this method needs to be overridden in \fIall\fP subclasses\&. This is typically done as part of setting up the full type system for extension classes, which is usually accomplished by using the pre-defined macros available through for instance \fBInventor/nodes/SoSubNode\&.h\fP (SO_NODE_INIT_CLASS and SO_NODE_CONSTRUCTOR for node classes), \fBInventor/engines/SoSubEngine\&.h\fP (for engine classes) and so on\&.
.PP
For more information on writing Coin extensions, see the class documentation of the toplevel superclasses for the various class groups\&. 
.PP
Implements \fBSoBase\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SS "const \fBSoFieldData\fP * SoFont::getFieldData (void) const\fC [protected]\fP, \fC [virtual]\fP"
Returns a pointer to the class-wide field data storage object for this instance\&. If no fields are present, returns \fCNULL\fP\&. 
.PP
Reimplemented from \fBSoFieldContainer\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SS "void SoFont::doAction (\fBSoAction\fP * action)\fC [virtual]\fP"
This function performs the typical operation of a node for any action\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SS "void SoFont::getBoundingBox (\fBSoGetBoundingBoxAction\fP * action)\fC [virtual]\fP"
Action method for the \fBSoGetBoundingBoxAction\fP\&.
.PP
Calculates bounding box and center coordinates for node and modifies the values of the \fIaction\fP to encompass the bounding box for this node and to shift the center point for the scene more towards the one for this node\&.
.PP
Nodes influencing how geometry nodes calculates their bounding box also overrides this method to change the relevant state variables\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SS "void SoFont::GLRender (\fBSoGLRenderAction\fP * action)\fC [virtual]\fP"
Action method for the \fBSoGLRenderAction\fP\&.
.PP
This is called during rendering traversals\&. Nodes influencing the rendering state in any way or who wants to throw geometry primitives at OpenGL overrides this method\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SS "void SoFont::callback (\fBSoCallbackAction\fP * action)\fC [virtual]\fP"
Action method for \fBSoCallbackAction\fP\&.
.PP
Simply updates the state according to how the node behaves for the render action, so the application programmer can use the \fBSoCallbackAction\fP for extracting information about the scene graph\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SS "void SoFont::pick (\fBSoPickAction\fP * action)\fC [virtual]\fP"
Action method for \fBSoPickAction\fP\&.
.PP
Does common processing for \fBSoPickAction\fP \fIaction\fP instances\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SS "void SoFont::getPrimitiveCount (\fBSoGetPrimitiveCountAction\fP * action)\fC [virtual]\fP"
Action method for the \fBSoGetPrimitiveCountAction\fP\&.
.PP
Calculates the number of triangle, line segment and point primitives for the node and adds these to the counters of the \fIaction\fP\&.
.PP
Nodes influencing how geometry nodes calculates their primitive count also overrides this method to change the relevant state variables\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.PP
Reimplemented in \fBSoFontStyle\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBSoSFName\fP SoFont::name"
Name of font\&.
.PP
Which fontnames are available is rather system dependent, not only on whether or not you are running on a UNIX/Linux system, Microsoft Windows or whatever, but also on which fonts and font \fItypes\fP (like TrueType) are installed on a particular user's system\&.
.PP
All font rendering nodes have a built-in fallback font to use, though, so even though Coin can not find a font on the system of the specified type, the text should be rendered somehow\&.
.PP
In summation, consider this node type and this particular field as a \fIhint\fP to the font rendering engines of Coin, and do \fInot\fP base your models on a particular font being available\&. 
.SS "\fBSoSFFloat\fP SoFont::size"
Size of font\&. Defaults to 10\&.0\&.
.PP
For 2D rendered bitmap fonts (like for \fBSoText2\fP), this value is the height of a character in screen pixels\&. For 3D text, this value is the world-space coordinates height of a character in the current units setting (see documentation for \fBSoUnits\fP node)\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


coin4/src/nodes/SoFont.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!  
  \class SoFont SoFont.h Inventor/nodes/SoFont.h
  \brief The SoFont class is an appearance node for setting fonts.
  \ingroup nodes

  Successive text rendering nodes (like SoText2, SoText3, SoAsciiText,
  etc) will use the font specified from an SoFont node when
  visualizing text.

  The font name is recognized with the form "family:style". The range of
  supported font families depends on which fonts are installed on the
  system. A typical font family might be "Arial" or "Times New
  Roman". 

  Style is either "Bold", "Italic" or "Bold Italic".

  Spaces before or after the ":" will be ignored.

  If a good match for the chosen font cannot be found, the default
  font will be loaded instead. The default 2D font is a builtin 8x12
  points font. The 3D font is a serif font ala Times New Roman. It is
  not possible to apply styles to any of the builtin fonts. It is not
  possible to specify a size for the default 2D font.
  
  One can explicitly select the default font by setting "defaultFont"
  as the font name, or by just leaving the SoFont::name field alone,
  as it has that value by default. In that regard, note that SoFont
  does not inherit / accumulate any parts from earlier SoFont-nodes,
  even when one or both of its fields are left blank in e.g. an
  iv-file, like this:

  \verbatim
  #Inventor V2.1 ascii
  
  Font { name "Helvetica"  size 50 }
  Text2 { string "hepp" }
  
  Translation { translation 5 80 0 }
  
  BaseColor { rgb 1 1 0 }
  
  Font { size 30 }
  Text2 { string "svupp" }
  \endverbatim

  (The second string in the above example will be rendered with the
  default built-in font, not in Helvetica.)


  The default fonts are always accessible by Coin as they are embedded
  into the run-time library. If one needs to guarantee that the text
  will have the same appearance under all circumstances, the default
  font will be a safe choice. Another solution might be to use the \e
  FreeType font engine and explicitly name a font-file. This is
  described in more detail below.

  Here is a simple example on how to print a string using a bold and
  italic Arial font:

  \verbatim
  #Inventor V2.1 ascii
  
  Separator {
     Font {
       name "Arial:Bold Italic"
       size 14
     }
     Text2 {
       string ["This is a", "Coin font test"]
       justification CENTER
     }    
  }
  \endverbatim

  Coin has support for two different font APIs. On non-Windows
  platforms, the \e FreeType library together with the optional \e
  Fontconfig library is used.

  If the \e Fontconfig library is installed the font file on the sytem
  for a given font name will be located through it. For more
  information on \e Fontconfig see
  http://freedesktop.org/software/fontconfig . Additionally \e
  Fontconfig allows to match specific fonts through its own pattern
  matching format (see
  http://pdx.freedesktop.org/~fontconfig/fontconfig-user.html).
  Please note that the point size value in the textual representation
  of the Fontconfig pattern is currently overriden by the size field
  of the SoFont node and if no size field is specified the default
  value of the size field is in effect. In case you intend to use your
  application on systems where the \e Fontconfig library is expected
  to be not installed you should not make use of \e Fontconfig's font
  pattern syntax. \e Fontconfig usage can be prevented by setting the
  "COIN_FORCE_FONTCONFIG_OFF" environment variable to 1.

  Here is an example on how to print a string using the \e Fontconfig
  pattern matching syntax using a bold & italic font where Times New
  Roman is the preferred font family.

  \verbatim
  #Inventor V2.1 ascii
  
  Separator {
     Font {
       name "Times New Roman,Arial,Helvetica:italic:bold"
       size 24
     }
     Text2 {
       string ["This is a", "Coin font test"]
       justification CENTER
     }    
  }
  \endverbatim


  On Windows the Win32 GDI library is used. \e FreeType and \e
  Fontconfig are dynamically loaded on demand by Coin if font support
  are requested by a node. When fontsupport is loaded on Windows,
  FreeType will have precedence over Win32 if located. This can be
  prevented by setting the "COIN_FORCE_FREETYPE_OFF" environment
  variable to 1. When using FreeType, you need FreeType version 2.1 or
  later. On Mac OS X, version 2.1.7 or later is required.

  If Coin cannot load the \e FreeType library, and is not running on
  Microsoft Windows, only the default fonts will be accessible.

  It is possible to specify the TrueType font file directly if \e
  FreeType is used as the font engine. This is done by including the
  ".ttf" in the filename, i.e. "Comic_Sans_MS.ttf". Coin will then
  search the local path for the running program and then the path
  specified by the "COIN_FONT_PATH" environment variable. If the
  program is using \e FreeType on a Windows platform, the
  "$WINDIR/Fonts" directory will also be searched.

  It is not possible to directly specify a TrueType font file if
  Windows is handling the fonts. This is due to the way Windows is
  accessing the fonts through the system registry. All fonts must
  therefore be properly installed and given a system name. Open the
  "Control Panel" and double click on the "Fonts" icon for an overview
  of installed fonts and their names.

  Beware that some non-English versions of Windows are using different
  name for the styles (i.e. "Italique" instead of "Italic"). These
  names are supported in Coin, but it is recommended for portability
  purposes to only use the English terms. Please note that there is
  still a possibility that there are no fonts installed using the
  terms "Bold" or "Italic" on the Windows platform. To guarantee that
  a font is accessible you must either use the \e FreeType library and
  include a TrueType font in your distribution, or you must avoid
  using styles and stick to the standard Windows fonts.

  If the "COIN_DEBUG_FONTSUPPORT" environment variable is set to 1, an
  extensive amount of information about loading, initializing and
  using fonts will be output. Issues like missing fonts and other
  related problems will then be reported, so we advice you to first
  try to use that debugging option when something does not work quite
  as expected.

  <b>FILE FORMAT/DEFAULTS:</b>
  \code
    Font {
        name "defaultFont"
        size 10
    }
  \endcode

  \sa SoFontStyle, SoText2, SoText3, SoAsciiText
*/

#include <Inventor/nodes/SoFont.h>

#include <string.h>

#include <Inventor/actions/SoCallbackAction.h>
#include <Inventor/actions/SoGLRenderAction.h>
#include <Inventor/actions/SoGetBoundingBoxAction.h>
#include <Inventor/actions/SoGetPrimitiveCountAction.h>
#include <Inventor/actions/SoPickAction.h>
#include <Inventor/elements/SoFontNameElement.h>
#include <Inventor/elements/SoFontSizeElement.h>
#include <Inventor/elements/SoOverrideElement.h>
#include <Inventor/errors/SoDebugError.h>

#include "nodes/SoSubNodeP.h"

// *************************************************************************

/*!
  \var SoSFName SoFont::name

  Name of font.

  Which fontnames are available is rather system dependent, not only
  on whether or not you are running on a UNIX/Linux system, Microsoft
  Windows or whatever, but also on which fonts and font \e types (like
  TrueType) are installed on a particular user's system.

  All font rendering nodes have a built-in fallback font to use,
  though, so even though Coin can not find a font on the system of the
  specified type, the text should be rendered somehow.

  In summation, consider this node type and this particular field as a
  \e hint to the font rendering engines of Coin, and do \e not base
  your models on a particular font being available.
*/

/*!
  \var SoSFFloat SoFont::size

  Size of font. Defaults to 10.0.

  For 2D rendered bitmap fonts (like for SoText2), this value is the
  height of a character in screen pixels. For 3D text, this value is
  the world-space coordinates height of a character in the current
  units setting (see documentation for SoUnits node).
*/

// *************************************************************************

SO_NODE_SOURCE(SoFont);

/*!
  Constructor.
*/
SoFont::SoFont(void)
{
  SO_NODE_INTERNAL_CONSTRUCTOR(SoFont);

  SO_NODE_ADD_FIELD(name, ("defaultFont"));
  SO_NODE_ADD_FIELD(size, (10.0f));
}

/*!
  Destructor.
*/
SoFont::~SoFont()
{
}

// Doc from superclass.
void
SoFont::initClass(void)
{
  SO_NODE_INTERNAL_INIT_CLASS(SoFont, SO_FROM_INVENTOR_1);

  SO_ENABLE(SoCallbackAction, SoFontNameElement);
  SO_ENABLE(SoCallbackAction, SoFontSizeElement);
  SO_ENABLE(SoGLRenderAction, SoFontNameElement);
  SO_ENABLE(SoGLRenderAction, SoFontSizeElement);
  SO_ENABLE(SoGetBoundingBoxAction, SoFontNameElement);
  SO_ENABLE(SoGetBoundingBoxAction, SoFontSizeElement);
  SO_ENABLE(SoGetPrimitiveCountAction, SoFontNameElement);
  SO_ENABLE(SoGetPrimitiveCountAction, SoFontSizeElement);
  SO_ENABLE(SoPickAction, SoFontNameElement);
  SO_ENABLE(SoPickAction, SoFontSizeElement);
}

// Doc from superclass.
void
SoFont::doAction(SoAction * action)
{
  SoState * state = action->getState();
  uint32_t flags = SoOverrideElement::getFlags(state);
    
#define TEST_OVERRIDE(bit) ((SoOverrideElement::bit & flags) != 0)
  
  if (!name.isIgnored() && !TEST_OVERRIDE(FONT_NAME)) {
    SoFontNameElement::set(state, this, this->name.getValue());
    if (this->isOverride()) {
      SoOverrideElement::setFontNameOverride(state, this, true);
    }
  }
  if (!size.isIgnored() && !TEST_OVERRIDE(FONT_SIZE)) {
    SoFontSizeElement::set(state, this, this->size.getValue());
    if (this->isOverride()) {
      SoOverrideElement::setFontSizeOverride(state, this, true);
    }
  }

#undef TEST_OVERRIDE
}

// Doc from superclass.
void
SoFont::getBoundingBox(SoGetBoundingBoxAction * action)
{
  SoFont::doAction(action);
}

// Doc from superclass.
void
SoFont::GLRender(SoGLRenderAction * action)
{
  SoFont::doAction(action);
}

// Doc from superclass.
void
SoFont::callback(SoCallbackAction * action)
{
  SoFont::doAction((SoAction *)action);
}

// Doc from superclass.
void
SoFont::pick(SoPickAction * action)
{
  SoFont::doAction(action);
}

// Doc from superclass.
void
SoFont::getPrimitiveCount(SoGetPrimitiveCountAction * action)
{
  SoFont::doAction(action);
}
